File: frontend/src/apis/published-content.ts
// SPDX-FileCopyrightText: 2025 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

import { apiClient } from './client'

// Published Content Request/Response Types
export interface PublishContentRequest {
  subtask_id: number
  content_type: 'video' | 'image'
  title: string
  description?: string
}

export interface PublishedWork {
  id: number
  user_id: number
  subtask_id: number
  attachment_id: number
  content_type: 'video' | 'image'
  title: string
  description: string | null
  prompt: string | null
  view_count: number
  like_count: number
  is_active: boolean
  is_featured: boolean
  created_at: string
  updated_at: string
  username: string | null
  avatar: string | null
  video_url: string | null
  video_thumbnail: string | null
  video_duration: number | null
  image_urls: string[] | null
}

export interface PublishedWorksListResponse {
  items: PublishedWork[]
  total: number
  page: number
  page_size: number
}

// Published Content API
export const publishedContentApi = {
  // Publish content to gallery
  publish: (data: PublishContentRequest) =>
    apiClient.post<PublishedWork>('/published-contents', data),

  // List published works (public, no auth)
  list: (params?: {
    content_type?: 'video' | 'image'
    page?: number
    page_size?: number
  }) => {
    const queryParams = new URLSearchParams()
    if (params?.content_type) queryParams.set('content_type', params.content_type)
    if (params?.page) queryParams.set('page', params.page.toString())
    if (params?.page_size) queryParams.set('page_size', params.page_size.toString())

    const queryString = queryParams.toString()
    const url = queryString ? `/published-contents?${queryString}` : '/published-contents'

    return apiClient.get<PublishedWorksListResponse>(url)
  },

  // Check if content is already published
  checkStatus: (subtaskId: number) =>
    apiClient.get<PublishedWork | null>(`/published-contents/check/${subtaskId}`),
}

File: backend/app/schemas/published_content.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class PublishedContentCreate(BaseModel):
    """Schema for creating a published content."""

    subtask_id: int = Field(..., description="ID of the subtask containing the generated content")
    content_type: str = Field(..., description="Type of content: video or image")
    title: str = Field(..., max_length=200, description="Title of the published work")
    description: Optional[str] = Field(
        None, max_length=2000, description="Description of the published work"
    )


class PublishedContentUpdate(BaseModel):
    """Schema for updating a published content."""

    title: Optional[str] = Field(None, max_length=200, description="New title")
    description: Optional[str] = Field(
        None, max_length=2000, description="New description"
    )


class PublishedContentResponse(BaseModel):
    """Schema for published content response with enriched data."""

    id: int
    user_id: int
    subtask_id: int
    attachment_id: int
    content_type: str
    title: str
    description: Optional[str] = None
    prompt: Optional[str] = None
    view_count: int
    like_count: int
    is_active: bool
    is_featured: bool
    created_at: datetime
    updated_at: datetime

    # Enriched data
    username: Optional[str] = Field(
        None, description="Username of the publisher (from User table)"
    )
    avatar: Optional[str] = Field(
        None, description="Avatar URL of the publisher (from User table)"
    )
    video_url: Optional[str] = Field(
        None, description="Video URL (from SubtaskContext.type_data)"
    )
    video_thumbnail: Optional[str] = Field(
        None, description="Video thumbnail URL (from SubtaskContext.type_data)"
    )
    video_duration: Optional[float] = Field(
        None, description="Video duration in seconds (from SubtaskContext.type_data)"
    )
    image_urls: Optional[List[str]] = Field(
        None, description="Image URLs (from SubtaskContext.type_data)"
    )

    class Config:
        from_attributes = True


class PublishedContentListResponse(BaseModel):
    """Schema for paginated published content list."""

    items: List[PublishedContentResponse]
    total: int
    page: int
    page_size: int
